Src/IronPython/Runtime/Operations/StringOps.cs

```diff
@@ -1169,51 +1169,68 @@ public static int rindex([NotNull]this string self, [NotNull]string sub, object?
                 }
             }
             return ret.ToString();
         }
 
-        public static string translate([NotNull]this string self, string? table) {
-            if (table == null) {
-                throw PythonOps.TypeError("'NoneType' object is not subscriptable");
-            } else if (table.Length != 256) {
-                throw PythonOps.ValueError("translation table must be 256 characters long");
-            } else if (self.Length == 0) {
+        private static void AppendValueForTranslate(this StringBuilder ret, object? mapped) {
+            switch (mapped) {
+                case null:
+                    return;
+                case int mappedInt:
+                    if (mappedInt > 0xFFFF) {
+                        throw PythonOps.TypeError("character mapping must be in range(0x10000)");
+                    }
+                    ret.Append((char)mappedInt);
+                    break;
+                case string _:
+                    ret.Append(mapped);
+                    break;
+                default:
+                    throw PythonOps.TypeError("character mapping must return integer, None or str");
+            }
+        }
+
+        public static string translate([NotNull] string self, [NotNull] string table) {
+            if (self.Length == 0) {
                 return self;
             }
 
-            // List<char> is about 2/3rds as expensive as StringBuilder appending individual
-            // char's so we use that instead of a StringBuilder
-            List<char> res = new List<char>();
+            StringBuilder ret = new StringBuilder();
             for (int i = 0; i < self.Length; i++) {
-                if (table != null) {
-                    int idx = (int)self[i];
-                    if (idx >= 0 && idx < 256) {
-                        res.Add(table[idx]);
-                    }
+                var idx = self[i];
+                if (idx < table.Length) {
+                    ret.Append(table[idx]);
                 } else {
-                    res.Add(self[i]);
+                    ret.Append(idx);
+                }
+            }
+            return ret.ToString();
+        }
+
+        public static string translate(CodeContext context, [NotNull] string self, object? table) {
+            if (self.Length == 0) {
+                return self;
+            }
+
+            if (!PythonTypeOps.TryGetOperator(context, table, "__getitem__", out object getitem)) {
+                throw PythonOps.TypeError($"'{PythonTypeOps.GetName(table)}' object is not subscriptable");
+            }
+
+            StringBuilder ret = new StringBuilder();
+            for (int i = 0; i < self.Length; i++) {
+                var idx = self[i];
+                try {
+                    ret.AppendValueForTranslate(PythonCalls.Call(context, getitem, (int)idx));
+                } catch (Exception e) when (e is LookupException || e is IndexOutOfRangeException || e is KeyNotFoundException) {
```

Comment:
I can't seem to think of a good way to check for `LookupError` from C# without checking every other derived type. @BCSharp have you run into anything like this before?

Reply:
I haven't, or better said, if I have, I have missed the complexity altogether.

Tn the code above, if the call throws a user-defined Python error that is a subclass of `LookupError`, `KeyError`, or `IndexError`, would the clause still catch it?

Reply:
Hmm, looks like user defined subclasses are not caught with the above. I guess I'd have to add an extra `|| PythonOps.IsInstance(e.GetPythonException(), PythonExceptions.LookupError)` or something like that.